src/realm/sync/network/default_socket.cpp

```diff
@@ -65,25 +65,95 @@ class DefaultWebSocketImpl final : public WebSocketInterface, public Config {
     void websocket_read_error_handler(std::error_code ec) override
     {
         m_logger.error("Reading failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::ReadError, ec.message()});
     }
     void websocket_write_error_handler(std::error_code ec) override
     {
         m_logger.error("Writing failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::WriteError, ec.message()});
     }
     void websocket_handshake_error_handler(std::error_code ec, const HTTPHeaders*,
                                            const std::string_view* body) override
     {
-        m_observer.websocket_handshake_error_handler(ec, body);
+        ErrorCodes::Error error;
+        bool was_clean = true;
+
+        if (ec == websocket::Error::bad_response_301_moved_permanently ||
+            ec == websocket::Error::bad_response_308_permanent_redirect) {
+            error = ErrorCodes::WebSocket_MovedPermanently;
+        }
+        else if (ec == websocket::Error::bad_response_3xx_redirection) {
+            error = ErrorCodes::Retry;
```

Comment:
Can we rename ErrorCodes::Retry to something more network or sync specific? The error codes in ErrorCodes are for the entire realm database and I don't really know what Retry means outside of the context of this websocket error handling function.

Reply:
I went with `WebSocket_Retry_Error`